File: hs/src/Day2.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PatternSynonyms #-}

module Day2 (part1, part2) where

import Control.Arrow (first, (***))
import Control.Monad (ap)
import Control.Monad.Fix (fix)
import Data.List (inits)
import Data.Text (Text)
import Data.Text qualified as T (splitOn, strip, pattern Empty, pattern (:<))
import Data.Text.Read qualified as T (decimal)

day2 :: (Integral a) => [(Int, a)] -> Text -> a
day2 factors = sum . concatMap go . T.splitOn ","
  where
    go line
      | Right (lo, '-' T.:< line') <- T.decimal $ T.strip line,
        Right (hi, T.Empty) <- T.decimal line' =
          [ (maxChunk * (maxChunk + 1) - minChunk * (minChunk + 1)) `div` 2 * (scale' * c)
          | (scale0, c) <- takeWhile ((<= hi) . fst) $ map (first $ (10 ^) . pred) factors,
            ((prev, _), (power, scale)) <-
              zip `ap` drop 1 $ takeWhile ((<= hi) . snd) $ iterate ((* 10) *** (* scale0)) (1, 1),
            let scale' = (scale * power - 1) `div` (power - 1)
                minChunk = max (prev - 1) $ (lo - 1) `div` scale'
                maxChunk = min (power - 1) $ hi `div` scale',
            minChunk < maxChunk
          ]
    go _ = []

part1, part2 :: Text -> Int
part1 = day2 [(2, 1)]
part2 = day2 $ filter ((/= 0) . snd) $ fix (zipWith go [2 ..] . inits)
  where
    go n factors = (n, foldl' (-) 1 [c | (m, c) <- factors, n `mod` m == 0])

File: rs/src/day2.rs
use std::sync::LazyLock;

fn day2(input: &str, factors: &[(u32, isize)]) -> isize {
    input
        .split(',')
        .filter_map(|range| {
            let (lo, hi) = range.trim().split_once('-')?;
            let lo = lo.parse::<isize>().ok()?;
            let hi = hi.parse::<isize>().ok()?;
            Some(
                factors
                    .iter()
                    .map_while(|(n, c)| {
                        let scale0 = 10isize.pow(*n - 1);
                        let mut power = 1;
                        let mut scale = 1;
                        let mut acc = 0;
                        loop {
                            let prev = power;
                            power *= 10isize;
                            scale *= scale0;
                            if scale > hi {
                                break;
                            }
                            let scale = (scale * power - 1) / (power - 1);
                            let lo = ((lo - 1) / scale).max(prev - 1);
                            let hi = (hi / scale).min(power - 1);
                            if lo < hi {
                                acc += (hi * (hi + 1) - lo * (lo + 1)) / 2 * scale * c;
                            }
                        }
                        Some(acc)
                    })
                    .sum::<isize>(),
            )
        })
        .sum::<isize>()
}

pub fn part1(input: &str) -> isize {
    day2(input, &[(2, 1)])
}

pub fn part2(input: &str) -> isize {
    static FACTORS: LazyLock<Vec<(u32, isize)>> = LazyLock::new(|| {
        let mut factors = vec![];
        for n in 2..=(isize::MAX as f64).log10() as u32 {
            let c = 1 - factors
                .iter()
                .filter_map(|(m, c)| if n % m == 0 { Some(*c) } else { None })
                .sum::<isize>();
            if c != 0 {
                factors.push((n, c));
            }
        }
        factors
    });
    day2(input, &FACTORS)
}
